mender-auth/client/client.cpp

```diff
@@ -0,0 +1,210 @@
+	return error::Error(
+		error_condition(code, AuthClientErrorCategory),
+		"Authentication error(" + resp->GetStatusMessage() + "): " + msg + "(" + response_body
+			+ ")");
+}
+
+error::Error GetJWTToken(
+	http::Client &client,
+	const string &server_url,
+	const string &private_key_path,
+	const string &device_identity_script_path,
+	events::EventLoop &loop,
+	APIResponseHandler api_handler,
+	const string &tenant_token,
+	const string &server_certificate_path) {
+	key_value_parser::ExpectedKeyValuesMap expected_identity_data =
+		identity_parser::GetIdentityData(device_identity_script_path);
+	if (!expected_identity_data) {
+		return expected_identity_data.error();
+	}
+	expected::ExpectedString expected_identity_data_json =
+		json::Dump(expected_identity_data.value());
+	if (!expected_identity_data_json) {
+		mlog::Error("Failed to dump the identity data to JSON");
+		return expected_identity_data_json.error();
+	}
+	auto identity_data_json = expected_identity_data_json.value();
+	mlog::Debug("Got identity data: " + identity_data_json);
+
+	// Create the request body
+	unordered_map<string, string> request_body_map {
+		{"id_data", identity_data_json},
+	};
+
+	if (tenant_token.size() > 0) {
+		request_body_map.insert({"tenant_token", tenant_token});
+	}
+
+	auto expected_public_key = crypto::ExtractPublicKey(private_key_path);
+	if (!expected_public_key) {
+		return expected_public_key.error();
+	}
+	request_body_map.insert({"pubkey", expected_public_key.value()});
+
+	auto expected_request_body = json::Dump(request_body_map);
+	if (!expected_request_body) {
+		return expected_request_body.error();
+	}
+	auto request_body = expected_request_body.value();
+
+	// Sign the body
+	auto expected_signature =
+		crypto::Sign(private_key_path, common::ByteVectorFromString(request_body));
+	if (!expected_signature) {
+		return expected_signature.error();
+	}
+	auto signature = expected_signature.value();
+
+	auto whole_url = path::Join(server_url, request_uri);
+
+	auto req = make_shared<http::OutgoingRequest>();
+	req->SetMethod(http::Method::POST);
+	req->SetAddress(whole_url);
+	req->SetHeader("Content-Type", "application/json");
+	req->SetHeader("Content-Length", to_string(request_body.size()));
+	req->SetHeader("Accept", "application/json");
+	req->SetHeader("X-MEN-Signature", signature);
+	req->SetHeader("Authorization", "API_KEY");
+
+	req->SetBodyGenerator([request_body]() -> io::ExpectedReaderPtr {
+		return make_shared<io::StringReader>(request_body);
+	});
+
+	auto received_body = make_shared<vector<uint8_t>>();
+
+	return client.AsyncCall(
+		req,
+		[received_body, api_handler](http::ExpectedIncomingResponsePtr exp_resp) {
+			if (!exp_resp) {
+				mlog::Error("Request failed: " + exp_resp.error().message);
+				api_handler(expected::unexpected(exp_resp.error()));
+				return;
+			}
+			auto resp = exp_resp.value();
+
+			auto body_writer = make_shared<io::ByteWriter>(received_body);
+			body_writer->SetUnlimited(true);
```

Comment:
@kacf A generic question (no need to change the code her) -- shouldn't this be basically forbidden and replaced by `received_body.resize(StringToLong(resp->GetHeader("Content-length")))`, ideally with a check of the actual size?

Reply:
Ideally yes. For now, it's good that we have to call this function, so that all locations can be easily searched. Later we may want to implement limits.